SIG-Debugging/2023/debugging-02-09.md

```diff
@@ -4,20 +4,132 @@
+* Nick:
+    * We spent time talking about that during the last meeting.
+    * We defined a cross-product according to the scenarios that we want to support.
+* Chris F:
+    * Meta question: what path are we going to take?
+    * One: design the entire framework?
+    * Two: design specific goals?
+    * Might be useful to define specific goals and that might help derive the discussion.
+* Jlbirch: 
+    * We need an end goal to be able to have ideas that fit together.
+    * General approach defining the entire framework and then implementing the pieces.
+* Nick:
+    * Many of the scenarios that we discussed last time have overlapping operations so we can take the common denominator of each and work from there. To avoid multiple implementations.
+    * Unification: inspecting multiple levels could be additive, depending on what you want to expect.
+* Ulrich:
+    * This ties back to the end user experience. 
+* Nick:
+    * We need to do some of the latter (guest debugging mostly) since it's what doesn't really exist. 
+* Ulrich:
+    * Part of this is we don't write our own debugger, we use existing debugging tools, right?
+* Till:
+    * Collect real use cases?
+    * Work from there backwards to inform where we start and have a clearer categorization.
+    * As a wasmtime developer… 
+    * As a JS developer… (debug JS code running in a VM running in Wasmtime).
+    * As a user compiling a Rust program…
+* Luke:
+    * Debugging the entire system (e.g via gdb) is very different than the other options, where you debug in different layers.
+* Chris F:
+    * We would have a simpler system if we don't generate DWARF.
+* Luke:
+    * Pick a protocol (e.g. Chrome Debug Protocol, LSP)
+    * Find an existing implementation that works natively
+    * Port to Wasm
+    * (Incremental approach to expose the Wasm state)
+* Chris F:
+    * +1 for the incremental approach.
+* Ralph:
+    * .NET implements multiple protocols
+    * They translate between multiple protocols
+    * The person who implemented it is available and could participate in this discussion.
+* Luke:
+    * Pick one or the other
+    * Put us in a better place
+* Ralph:
+    * The Chrome protocol is not strictly open source; the team assumes that GDB is "canonical".
+* Till:
+    * Implications between components when choosing a protocol.
+* Luke:
+    * If protocol needs a stack trace, we don't know what happens if we switch languages (multiple components, multiple languages).
+* Nick:
+    * Assumptions of a single memory for example.
+* Ulrich:
+    * LLDB is pretty much equivalent to GDB
+    * GDB itself constructs high level information from the low-level information (registers, memory)
+* Chris F: 
+    * Browser level seems a better fit, for the semantics of Wasm. We don't want to expose the machine internals (not necessarily).
```

Comment:
I think using the gdb/lldb debugging protocols makes more sense than the browser debugging protocols. The browser debugging protocols assume that the given javascript (or other language) is what the user wrote and has high level types. Gdb/lldb on the other hand assume that the executed language (in this case webassembly) is compiled from a higher level language (like rust) and provide the ability to inspect both the executed language and the higher level language. If you were to do the reconstruction in wasmtime to allow using a browser debugging protocols, it would become very hard to debug a program that has memory corruption, as you are only seeing the higher level view without any ability to know that `var_b` was stored on the stack and overwritten by an out of bounds write of `buf_a`. If on the other hand wasmtime does no reconstruction, the user would see just raw bytes that are useless to them.

As for handling of components. Gdb and lldb should handle mixed language processes just fine. As for the single memory assumption, you could present a 64bit address space and map the 32bit memories of each component at a different location I think.

Reply:
As I recall this conversation, someone had raised that browser devtools might be able to interpret Wasm DWARF; no one was suggesting that we adopt a JS-only debugger (that indeed wouldn't make sense in a Wasm context). The broader question was about the abstraction level in the protocol: does the debugger interface with debuggee at the native-code level (and thus require mappings from Wasm PCs and Wasm locals to native PCs and registers/stack), or at the Wasm bytecode level? Unless one also needs to debug the JIT'd code, the latter is likely to be more robust, as it's one less level of translation. It's also required if one is debugging a Wasm producer -- the level one really wants is the Wasm VM state, which is unavailable if one only has a native-level debugger and doesn't yet emit DWARF. (The case where one debugs JIT'd code is also useful and we want to support that too, so maybe we'll have two approaches eventually.) We had an impression that browser debug protocols were closer to the latter, which is why they came up. In contrast, this would imply a port of gdb or lldb to a new target (wasm32).

A meta point too: you'd be welcome I'm sure to join the SIG if you're interested (@fitzgen is its chair); that's probably better for these sorts of conversations than a post-hoc thread on the notes :-)